flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/StandaloneCompletedCheckpointStoreTest.java

```diff
@@ -116,4 +120,35 @@ public void testAddCheckpointWithFailedRemove() throws Exception {
 			}
 		}
 	}
+
+	@Test
+	public void preferCheckpointForRecoveryWithSavepoint() throws Exception {
+		StandaloneCompletedCheckpointStore store = new StandaloneCompletedCheckpointStore(5);
+		JobID jobId = new JobID();
+		store.addCheckpoint(checkpoint(jobId, 1L));
+		store.addCheckpoint(checkpoint(jobId, 2L));
+		store.addCheckpoint(checkpoint(jobId, 3L));
```

Comment:
The `savepoint` method is not used so far. Maybe it should be used by an additional test like:
```
checkpoint 1, checkpoint 2, savepoint 3 -> returns checkpoint 2
```